### PR TITLE
[tanjiro] Laplacian curvature (κ_H, κ_G) as surface input features (Wave 21)

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

High-curvature surface regions (sharp edges, wheel arches, A-pillars) are where wall-shear stress (wsy, wsz) errors cluster in CFD surrogates. These regions concentrate boundary-layer separation and reattachment events that are poorly captured without explicit local geometry cues.

We hypothesize that adding **Laplacian-derived surface features** (mean curvature κ_H and Gaussian curvature κ_G computed from the mesh) as additional input channels will directly inform the model about the geometry regions it struggles with most, reducing the dominant wsy/wsz errors.

Motivation: Discrete Laplacian operators on triangle meshes provide a principled, coordinate-invariant way to compute local curvature scalars. High-|κ_H| regions correspond exactly to the aerodynamic features driving wsy/wsz errors (separation lines, stagnation zones). Related work: point cloud normal estimation (PCPNet, 2018) and physics-informed mesh neural nets consistently show curvature features improve surface quantity prediction.

## Instructions

Implement in `target/train.py`. Add two scalar surface features computed from the mesh:

1. **Mean curvature κ_H** per surface vertex: derived from cotangent Laplacian applied to vertex positions. Approximate as: `κ_H ≈ (L_cot @ positions).norm(dim=1)` where L_cot is the cotangent weight matrix.
2. **Gaussian curvature κ_G** per surface vertex: compute via angle-defect formula: `2π - sum(incident_triangle_angles)`.

These two scalars are concatenated to the existing per-surface-point input features before the linear input projection.

**Concrete steps:**

```python
# In the dataset/data loader, for each surface mesh:
# 1. Build cotangent Laplacian L_cot from face connectivity
# 2. mean_curv = ||L_cot @ verts||  (shape: [N_verts])
# 3. gauss_curv = angle_defect per vertex  (shape: [N_verts])
# 4. Append as features: input_feats = torch.cat([existing_feats, mean_curv.unsqueeze(-1), gauss_curv.unsqueeze(-1)], dim=-1)
```

Normalize both curvature values (zero-mean, unit-variance over the dataset training split) before feeding into the model.

The curvature values are **surface-only** (volume points don't have mesh curvature), so only the surface branch input projection layer changes. Volume input dim remains unchanged.

Standard training flags:
```bash
torchrun --standalone --nproc-per-node=4 train.py \
  --model-layers 5 --model-hidden-dim 256 --model-heads 4 \
  --fourier-pe --no-use-ema \
  --lr 3e-4 --lr-cosine-t-max 50 \
  --no-compile-model \
  --wandb-group bengio-wave21-tanjiro-laplacian-curv
```

Report per-axis breakdown (wsy, wsz separately) in your results — the key question is whether the curvature signal specifically improves the high-error wall-shear axes.

## Kill gates (abupt)
- EP5: kill if > 13%
- EP10: kill if > 10.5%
- EP15: kill if > 8.5%
- EP20: kill if > 8.0%
- EP25: kill if ≥ 6.9549% (current best baseline)

## Baseline

Current best: **PR #174** (alphonse, 5L/256d/4H + FourierEmbed + T_max=50, no-EMA)
W&B run: `vu4jsiic`, ep~45.3, step 807,025

| Metric | Current Best (val) | AB-UPT Target |
|--------|-------------------|---------------|
| `val_primary/abupt_axis_mean_rel_l2_pct` | **6.9549%** | 4.51% |
| `val_primary/surface_pressure_rel_l2_pct` | 4.5644% | 3.82% |
| `val_primary/volume_pressure_rel_l2_pct` | **3.9361%** ✓ | 6.08% |
| `val_primary/wall_shear_y_rel_l2_pct` | 8.7345% | 3.65% |
| `val_primary/wall_shear_z_rel_l2_pct` | 10.5766% | 3.63% |

Binding constraint: wsy (5.08pp gap) and wsz (6.95pp gap).

Reproduce command (baseline):
```bash
cd target/ && torchrun --standalone --nproc-per-node=4 train.py \
  --model-layers 5 --model-hidden-dim 256 --model-heads 4 \
  --fourier-pe --no-use-ema \
  --lr 3e-4 --lr-cosine-t-max 50 \
  --no-compile-model
```

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT tanjiro:

## Pre-implementation finding — pre-computed `curvature_HK_v2.npy` already on PVC

While exploring the data layout I discovered that **mesh curvature is already pre-computed** and stored in each case directory as `curvature_HK_v2.npy`. Quick summary:

- File: `<case>/curvature_HK_v2.npy`, shape `(N_surface, 2)`, float32
- Col 0 = mean curvature κ_H (mean=-13.6, std=113, p1/p99=[-427, 264])
- Col 1 = Gaussian curvature κ_G (mean=-259, std=19506, p1/p99=[-8594, 7093])
- Sign of κ_H matches what we'd expect from `(L_cot @ verts) · normal` (signed mean curvature)
- Magnitudes are consistent with mesh-resolved CFD geometry (sharp edges → large κ)

**Coverage gap**: 62 of 484 cases are missing this file (57 train, 4 val, 1 test). Files were created today (May 3 06:57–07:40) — looks like a partial generation run.

**Critical**: The PR instructions ask to compute κ_H/κ_G via cotangent Laplacian and angle-defect, but the dataset on disk only has point clouds (no triangle face connectivity exposed by `data/loader.py`). The pre-computed file appears to bypass this issue (likely computed from the original `.vtu` meshes by an upstream process).

**Proposed plan** (would appreciate your sign-off):
1. Use the pre-computed `curvature_HK_v2.npy` as the source of κ_H, κ_G (matches the spirit of the hypothesis)
2. For the 62 missing cases: zero-fill curvature features (post-standardization → equals population mean). Add binary `has_curv` flag (0/1) as a 3rd extra channel so the model can learn the imputation pattern. **Or** drop missing cases entirely (would lose 14% train, 12% val — risky).
3. Apply `signed_log1p(x) = sign(x) * log(1+|x|)` to handle the heavy tails (col 0 → range [-7.6, 7.2]; col 1 → [-15.4, 14.6])
4. Standardize using train-split mean/std of the log-transformed features
5. Concatenate curvature features to surface_x (dim 7 → 9 with zero-fill flag, or 7 → 9 plus 1 mask = 10)
6. Update `surface_input_dim` in model build

**Questions for you**:
- (a) Use the pre-computed file as proposed, or do you want me to add a from-scratch computation in the dataset (would need to re-build a graph/kNN since faces aren't exposed)?
- (b) For 62 missing cases: zero-fill+mask flag, or drop?
- (c) Are the kill gates (EP5 ≤13%, EP10 ≤10.5%) still appropriate given this changes the input dimensionality but not architecturally novel?

I'll wait briefly for your reply before kicking off the run. If no reply within ~30 min I'll proceed with the proposed plan (pre-computed + zero-fill+mask flag + signed-log standardization).

**Comment by morganmcg1** [comment]
ADVISOR: Great pre-implementation audit — using the pre-computed `curvature_HK_v2.npy` is exactly the right call, no need to re-derive from scratch.

**Answers to your questions:**

**(a) Pre-computed file:** Yes, use it. The hypothesis tests whether curvature signal helps, not whether we can compute curvature ourselves. The pre-computed values from the original `.vtu` meshes are likely higher quality than anything we'd get from a point-cloud kNN approximation anyway.

**(b) Missing cases (62/484):** Use **zero-fill + mask flag** (your option 2). Do NOT drop cases — losing 14% of training data is too costly. The mask flag is smart: the model can learn to ignore curvature on zero-fill cases. Implement as: append a single binary `has_curv` scalar (1.0 if file present, 0.0 if missing) to get surface_input_dim = 10 (7 existing + κ_H + κ_G + has_curv).

**(c) Kill gates:** The same gates apply. Input dimensionality changes don't affect gate logic — we're still measuring the same abupt metric.

**Standardization:** Your signed-log1p + train-split mean/std is the right approach. Compute train-split stats once at dataset init and store them; don't recompute per batch.

**One important detail:** When you load the pre-computed file, verify the surface point ordering matches the ordering of surface points in the main point cloud (i.e., same N_surface, same vertex order). If it mismatches you'll silently corrupt features. Add an assert that the curvature array shape[0] == surface_xyz.shape[0] for each case.

Proceed immediately. Standard flags as specified in the PR body. Launch now.

**Comment by morganmcg1** [comment]
ADVISOR (escalation, 2026-05-03): Implementation questions answered at 09:16Z — please confirm you have launched your training run with the pre-computed `curvature_HK_v2.npy` approach and post your W&B run ID(s). The baseline to beat is **val_abupt = 6.9549%** (PR #174, run `vu4jsiic`).

Post ep1 metrics as soon as available and apply standard gates (ep5: kill if abupt ≥ 13%, ep10: ≥ 10.5%, ep15: ≥ 8.5%).

**Comment by morganmcg1** [comment]
STUDENT tanjiro:

## Training launched ✅

Implementation complete on the branch (commit pending — will push with results). Training started at 09:43Z on 2026-05-03.

**W&B run (rank0):** [`kaemek2z`](https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml/runs/kaemek2z) — `tanjiro/laplacian-curv-v1`
- group: `bengio-wave21-tanjiro-laplacian-curv`
- other ranks: `kaemek2z` (r0), `x5bvak1s` (r1), `5ha9949l` (r2), `2e4kyw3e` (r3)

**Implementation summary** (matches your sign-off at 09:16Z):
- Pre-computed `curvature_HK_v2.npy` (κ_H, κ_G), zero-fill + binary `has_curv` mask flag for the 62 missing cases
- `signed_log1p(x) = sign(x)·log(1+|x|)` then standardize using train-split mean/std (computed once at init from 20 cases × 200k pts each)
- Optional ±5σ clip on the standardized values
- Surface input dim: 7 → 10 (κ_H, κ_G, has_curv); volume input untouched
- Asserts curvature `shape[0] == n_surface_full` per case
- Gated by `--use-curvature` so the codepath is opt-in (baseline preserved)

**Launch command:**
```bash
cd target/ && torchrun --standalone --nproc-per-node=4 train.py \
  --use-curvature \
  --model-layers 5 --model-hidden-dim 256 --model-heads 4 \
  --fourier-pe --no-use-ema \
  --lr 3e-4 --lr-cosine-t-max 50 \
  --no-compile-model \
  --wandb-group bengio-wave21-tanjiro-laplacian-curv \
  --wandb-name tanjiro/laplacian-curv-v1 \
  --epochs 50 \
  --kill-thresholds 80000:val_primary/abupt_axis_mean_rel_l2_pct<13.0,170000:val_primary/abupt_axis_mean_rel_l2_pct<10.5,260000:val_primary/abupt_axis_mean_rel_l2_pct<8.5,350000:val_primary/abupt_axis_mean_rel_l2_pct<8.0,440000:val_primary/abupt_axis_mean_rel_l2_pct<6.9549
```

**Progress (as of ~09:51Z):** Epoch 1/50, ~step 4060, ~8.8 it/s. First validation will land at ep1 end (~step ~17816). Will post ep1/ep5/etc. metrics as they appear and let the in-train kill thresholds gate.
